# Import Libraries and Data

In [24]:
# import libraries

import pandas as pd
import numpy as np
import seaborn as sns
import os

In [25]:
# define path
path = r'/Users/mimielaurant/Documents/Career Foundry/Immersion_Achievement 6/Citibike Project'

In [26]:
# import citibike.csv

df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'citibike.csv'), index_col = False)

In [27]:
df.head()

trip_id  bike_id weekday  start_hour           start_time  start_station_id  \
0  LnQzQk    16013     Mon          18  2013-09-09 18:18:55               523   
1  IL9boN    15230     Thu          18  2013-09-12 18:38:53               257   
2  46clGB    17942     Wed          19  2013-09-18 19:44:04               479   
3  v7vdFt    19683     Sat          11  2013-09-28 11:54:37               527   
4  VGBsb5    18024     Sat          18  2013-09-07 18:08:22               521   

        start_station_name  start_station_latitude  start_station_longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

              end_time  end_station_id     end_station_name  \
0  2013-09-09 18:35:28             334      W 20 St & 7 Ave   
1  2013-09-12 18:48:34             236  St Marks Pl & 2 Ave   
2  2013-09-18 19:50:05             513     W 56 St & 10 Ave   
3  2013-09-28 12:03:58             441      E 52 St & 2 Ave   
4  2013-09-07 18:46:38             476      E 31 St & 3 Ave   

   end_station_latitude  end_station_longitude  trip_duration      subscriber  \
0             40.742388             -73.997262            993      Subscriber   
1             40.728419             -73.987140            581      Subscriber   
2             40.768254             -73.988639            361      Subscriber   
3             40.756014             -73.967416            561      Subscriber   
4             40.743943             -73.979661           2296  Non-Subscriber   

   birth_year  gender  
0      1968.0       2  
1      1983.0       1  
2      1989.0       1  
3      1988.0       2  
4         NaN       0

In [28]:
df.shape

(50000, 18)

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   trip_id                  50000 non-null  object 
 1   bike_id                  50000 non-null  int64  
 2   weekday                  50000 non-null  object 
 3   start_hour               50000 non-null  int64  
 4   start_time               50000 non-null  object 
 5   start_station_id         50000 non-null  int64  
 6   start_station_name       50000 non-null  object 
 7   start_station_latitude   50000 non-null  float64
 8   start_station_longitude  50000 non-null  float64
 9   end_time                 50000 non-null  object 
 10  end_station_id           50000 non-null  int64  
 11  end_station_name         50000 non-null  object 
 12  end_station_latitude     50000 non-null  float64
 13  end_station_longitude    50000 non-null  float64
 14  trip_duration         

In [30]:
df.describe()

bike_id    start_hour  start_station_id  start_station_latitude  \
count  50000.000000  50000.000000      50000.000000            50000.000000   
mean   17615.269360     14.145240        443.321500               40.734170   
std     1675.407446      4.860541        356.559925                0.019911   
min    14556.000000      0.000000         72.000000               40.680342   
25%    16188.000000     10.000000        304.000000               40.720196   
50%    17584.000000     15.000000        402.000000               40.735877   
75%    19014.000000     18.000000        484.000000               40.750020   
max    20642.000000     23.000000       3002.000000               40.770513   

       start_station_longitude  end_station_id  end_station_latitude  \
count             50000.000000    50000.000000          50000.000000   
mean                -73.991109      442.539700             40.733859   
std                   0.012555      355.756022              0.019885   
min                 -74.017134       72.000000             40.680342   
25%                 -74.000271      304.000000             40.720196   
50%                 -73.990765      402.000000             40.735354   
75%                 -73.981923      483.000000             40.749013   
max                 -73.950048     3002.000000             40.770513   

       end_station_longitude  trip_duration    birth_year        gender  
count           50000.000000   50000.000000  43021.000000  50000.000000  
mean              -73.991351     838.982900   1975.627786      1.073540  
std                 0.012569     573.663997     11.089001      0.589389  
min               -74.017134      60.000000   1899.000000      0.000000  
25%               -74.001547     417.000000   1968.000000      1.000000  
50%               -73.991218     672.000000   1978.000000      1.000000  
75%               -73.982050    1112.000000   1984.000000      1.000000  
max               -73.950048    2697.000000   1997.000000      2.000000

# Consistency Checks

In [31]:
# check for mixed-type columns

for col in df.columns.tolist():
  weird = (df [[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

In [32]:
# check for missing values

df.isnull().sum()

trip_id                       0
bike_id                       0
weekday                       0
start_hour                    0
start_time                    0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_time                      0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
trip_duration                 0
subscriber                    0
birth_year                 6979
gender                        0
dtype: int64

In [33]:
df.median()

/var/folders/mc/x6vs3r5x383gsf2kmlm681sh0000gn/T/ipykernel_91963/530051474.py:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.median()


bike_id                    17584.000000
start_hour                    15.000000
start_station_id             402.000000
start_station_latitude        40.735877
start_station_longitude      -73.990765
end_station_id               402.000000
end_station_latitude          40.735354
end_station_longitude        -73.991218
trip_duration                672.000000
birth_year                  1978.000000
gender                         1.000000
dtype: float64

In [34]:
# create subset of data with missing birth years

df_nan = df[df['birth_year'].isnull() == True]

In [35]:
df_nan

trip_id  bike_id weekday  start_hour           start_time  \
4      VGBsb5    18024     Sat          18  2013-09-07 18:08:22   
10     aZsCdK    15732     Sat          20  2013-09-07 20:01:58   
17     VqhF3S    15651     Mon          18  2013-09-30 18:32:48   
26     9yZrq6    17506     Sun          16  2013-09-01 16:18:58   
46     t1KBiF    19206     Sun          22  2013-09-01 22:49:52   
...       ...      ...     ...         ...                  ...   
49986  8wDcq8    19293     Sun          22  2013-09-29 22:00:43   
49988  C0ivmQ    18118     Sun          12  2013-09-22 12:19:44   
49993  cvgHF4    18458     Wed          19  2013-09-25 19:01:07   
49995  Fb2bmC    17988     Sat          18  2013-09-21 18:39:20   
49998  xxacSt    16763     Sat          13  2013-09-07 13:25:35   

       start_station_id      start_station_name  start_station_latitude  \
4                   521         8 Ave & W 31 St               40.750450   
10                  457      Broadway & W 58 St               40.766953   
17                  363          West Thames St               40.708347   
26                  389     Broadway & Berry St               40.710446   
46                  426   West St & Chambers St               40.717548   
...                 ...                     ...                     ...   
49986               469      Broadway & W 53 St               40.763441   
49988               352         W 56 St & 6 Ave               40.763406   
49993               284   Greenwich Ave & 8 Ave               40.739017   
49995               479         9 Ave & W 45 St               40.760193   
49998               271  Ashland Pl & Hanson Pl               40.685282   

       start_station_longitude             end_time  end_station_id  \
4                   -73.994811  2013-09-07 18:46:38             476   
10                  -73.981693  2013-09-07 20:23:51             512   
17                  -74.017134  2013-09-30 19:01:59             514   
26                  -73.965251  2013-09-01 16:26:56             389   
46                  -74.013221  2013-09-01 23:06:48             247   
...                        ...                  ...             ...   
49986               -73.982681  2013-09-29 22:17:52             444   
49988               -73.977225  2013-09-22 12:29:55             464   
49993               -74.002638  2013-09-25 19:19:11             329   
49995               -73.991255  2013-09-21 20:08:59             478   
49998               -73.978058  2013-09-07 14:14:04             373   

                   end_station_name  end_station_latitude  \
4                   E 31 St & 3 Ave             40.743943   
10                  W 29 St & 9 Ave             40.750073   
17                 12 Ave & W 40 St             40.760875   
26              Broadway & Berry St             40.710446   
46           Perry St & Bleecker St             40.735354   
...                             ...                   ...   
49986            Broadway & W 24 St             40.742354   
49988               E 56 St & 3 Ave             40.759345   
49993     Greenwich St & N Moore St             40.720434   
49995              11 Ave & W 41 St             40.760301   
49998  Willoughby Ave & Walworth St             40.693317   

       end_station_longitude  trip_duration      subscriber  birth_year  \
4                 -73.979661           2296  Non-Subscriber         NaN   
10                -73.998393           1313  Non-Subscriber         NaN   
17                -74.002777           1751  Non-Subscriber         NaN   
26                -73.965251            478  Non-Subscriber         NaN   
46                -74.004831           1016  Non-Subscriber         NaN   
...                      ...            ...             ...         ...   
49986             -73.989151           1029  Non-Subscriber         NaN   
49988             -73.967597            611  Non-Subscriber         NaN   
49993             -74.010206           

In [36]:
# impute median value into missing birth years 

df['birth_year'] = df['birth_year'].fillna(df['birth_year'].median())


I chose to impute the median instead of mean to try to account for any outliers, and to reduce the risk of skewing the data. I did not want to delete 13% of the data and risk further analysis being drastically altered.

In [37]:
# second check for missing values

df.isnull().sum()

trip_id                    0
bike_id                    0
weekday                    0
start_hour                 0
start_time                 0
start_station_id           0
start_station_name         0
start_station_latitude     0
start_station_longitude    0
end_time                   0
end_station_id             0
end_station_name           0
end_station_latitude       0
end_station_longitude      0
trip_duration              0
subscriber                 0
birth_year                 0
gender                     0
dtype: int64

In [38]:
# check for duplicates
df_dups = df[df.duplicated()]

In the descriptive stats above, it says that the minimum birth year is 1899, I highly doubt that someone born in 1899 woul dbe renting a bike in 2013. I'm going to check the birth year colun for extreme values.

In [39]:
# check for birth years prior to 1913

df[df['birth_year'] <1913]

trip_id  bike_id weekday  start_hour           start_time  \
563    wkmOxr    19507     Mon          12  2013-09-23 12:43:58   
1371   M2EaCC    16277     Thu          11  2013-09-05 11:20:55   
1431   j2Vg6q    14638     Thu           7  2013-09-26 07:35:18   
1754   BQLrTb    17594     Wed          17  2013-09-25 17:00:04   
4823   S6IfuK    19804     Thu           7  2013-09-12 07:36:39   
7502   qawZh0    17490     Mon           7  2013-09-30 07:38:46   
9644   wBkIk3    15732     Sun          11  2013-09-08 11:11:35   
12926  xokid3    14944     Fri          20  2013-09-20 20:16:18   
13081  e65yHf    20512     Tue           7  2013-09-10 07:53:12   
13575  UJBxGg    18462     Wed           8  2013-09-11 08:40:59   
16135  ZQu9R0    15685     Tue          18  2013-09-10 18:53:53   
16252  Q4Ptz2    14581     Thu          17  2013-09-26 17:48:15   
17130  DGkBPe    19744     Thu           7  2013-09-12 07:12:03   
18332  Ax7tdR    19746     Wed          11  2013-09-11 11:21:53   
19824  ZjJvz9    15104     Thu          18  2013-09-05 18:11:52   
28568  GeuiSj    18769     Sun          11  2013-09-22 11:07:36   
37638  PvFonw    19667     Thu          19  2013-09-05 19:42:45   
39863  rDtB8y    19775     Sun          13  2013-09-08 13:37:24   
40359  Hy6IKm    19749     Thu           7  2013-09-19 07:58:17   
44309  l6SHHv    15521     Wed          16  2013-09-11 16:22:35   
47919  eGeYyS    19507     Sun          11  2013-09-22 11:20:44   
48989  RV9zZM    16098     Tue          17  2013-09-03 17:44:39   
49684  52CNru    15708     Mon          14  2013-09-30 14:54:49   

       start_station_id             start_station_name  \
563                 305                E 58 St & 3 Ave   
1371                500             Broadway & W 51 St   
1431                427                       State St   
1754                476                E 31 St & 3 Ave   
4823                515               W 43 St & 10 Ave   
7502                498             Broadway & W 32 St   
9644                515               W 43 St & 10 Ave   
12926               427                       State St   
13081               471        Grand St & Havemeyer St   
13575               514               12 Ave & W 40 St   
16135               268          Howard St & Centre St   
16252               358  Christopher St & Greenwich St   
17130               427                       State St   
18332               327       Vesey Pl & River Terrace   
19824               540        Lexington Ave & E 26 St   
28568              2004              6 Ave & Broome St   
37638               306           Cliff St & Fulton St   
39863               327       Vesey Pl & River Terrace   
40359               147       Greenwich St & Warren St   
44309               225        W 14 St & The High Line   
47919               347       W Houston St & Hudson St   
48989               485                W 37 St & 5 Ave   
49684               328        Watts St & Greenwich St   

       start_station_latitude  start_station_longitude             end_time  \
563                 40.760958               -73.967245  2013-09-23 12:52:50   
1371                40.762288               -73.983362  2013-09-05 11:36:10   
1431                40.702515               -74.014270  2013-09-26 08:00:02   
1754                40.743943               -73.979661  2013-09-25 17:09:33   
4823                40.760094               -73.994618  2013-09-12 07:55:02   
7502                40.748549               -73.988084  2013-09-30 08:01:17   
9644                40.760094               -73.994618  2013-09-08 11:32:02   
12926               40.702515               -74.014270  2013-09-20 20:58:45   
13081               40.712868               -73.956981  2013-09-10 08:17:33   
13575               40.760875               -74.002777  2013-09-11 08:59:17   
16135               40.719105               -73.999733  2013-09-10 19:14:46   
16252               40.732916               -74.007114  

Since there are less than 30 records with birth years prior to 1913, I will delete them from the dataset. Choosing 1913 was generous, however there are some spry centenarians out there!

In [40]:
df_new = df[df['birth_year'] > 1913]

In [41]:
df_new.shape

(49977, 18)

In [42]:
df_new.head()

trip_id  bike_id weekday  start_hour           start_time  start_station_id  \
0  LnQzQk    16013     Mon          18  2013-09-09 18:18:55               523   
1  IL9boN    15230     Thu          18  2013-09-12 18:38:53               257   
2  46clGB    17942     Wed          19  2013-09-18 19:44:04               479   
3  v7vdFt    19683     Sat          11  2013-09-28 11:54:37               527   
4  VGBsb5    18024     Sat          18  2013-09-07 18:08:22               521   

        start_station_name  start_station_latitude  start_station_longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

              end_time  end_station_id     end_station_name  \
0  2013-09-09 18:35:28             334      W 20 St & 7 Ave   
1  2013-09-12 18:48:34             236  St Marks Pl & 2 Ave   
2  2013-09-18 19:50:05             513     W 56 St & 10 Ave   
3  2013-09-28 12:03:58             441      E 52 St & 2 Ave   
4  2013-09-07 18:46:38             476      E 31 St & 3 Ave   

   end_station_latitude  end_station_longitude  trip_duration      subscriber  \
0             40.742388             -73.997262            993      Subscriber   
1             40.728419             -73.987140            581      Subscriber   
2             40.768254             -73.988639            361      Subscriber   
3             40.756014             -73.967416            561      Subscriber   
4             40.743943             -73.979661           2296  Non-Subscriber   

   birth_year  gender  
0      1968.0       2  
1      1983.0       1  
2      1989.0       1  
3      1988.0       2  
4      1978.0       0

# Export Data

In [43]:
df_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'citibikes_clean.csv'))